File: banc-essai/paysage.js
/* ===================== BANC « PAYSAGE » =====================
   Ce qu'on vérifie : téléphone couché, on ne doit RIEN voir de l'app.
   Le voile « Tourne l'écran » est un fond plein, mais depuis qu'il arrive
   avec l'horloge du jeu (0,52 s) il y a une demi-seconde où l'on voit au
   travers. Et ce qu'on voyait, c'était la colonne figée à 393 px collée à
   gauche d'un écran de 852 : l'écran coupé en deux.
   Trois mesures, sur trois téléphones :
     1. le voile est levé ET l'app est invisible (les deux conditions du
        style doivent tomber ensemble) ;
     2. à voile transparent, aucune couture verticale au bord de la colonne
        (on compare avec la plus forte couture ailleurs dans la photo, qui
        sert de témoin) ;
     3. redressé, l'app est de nouveau visible tout de suite.
   Usage : node banc-essai/paysage.js [url]
*/
const { chromium } = require('/opt/node22/lib/node_modules/playwright');
const { execSync } = require('child_process');
const fs = require('fs');

const URL = process.argv[2] || 'http://127.0.0.1:8099/index.html';
const D = '/tmp/claude-0/-home-user-BibleTrivia/fb9bf869-826b-5523-9825-ea1b24c294d0/scratchpad/';
const S = 2;                       /* échelle : 2 px physiques par px CSS */

const TELS = [
  ['iPhone 15',      393, 852],
  ['Pixel / Xiaomi', 412, 915],
  ['grand pliant',   500, 1023],
];

(async () => {
  const nav = await chromium.launch({ executablePath: '/opt/pw-browsers/chromium' });
  let ok = true;
  console.log('  téléphone         voile   app couchée   couture au bord   témoin ailleurs   retour');
  for (const [nom, w, h] of TELS) {
    const ctx = await nav.newContext({ viewport:{ width:w, height:h }, deviceScaleFactor:S,
      isMobile:true, hasTouch:true, serviceWorkers:'block' });
    const p = await ctx.newPage();
    await p.goto(URL, { waitUntil:'networkidle' });
    await p.waitForTimeout(1200);

    await p.setViewportSize({ width:h, height:w });   /* on couche le téléphone */
    await p.waitForTimeout(150);

    const e = await p.evaluate(() => {
      const rl = document.getElementById('rotate-lock');
      const app = document.getElementById('app');
      return {
        voile: getComputedStyle(rl).display !== 'none',
        vis:   getComputedStyle(app).visibility,
        bord:  app.getBoundingClientRect().right,
      };
    });

    /* Voile rendu transparent : on regarde exactement ce que l'oeil voit à
       la première image de son arrivée. */
    await p.evaluate(() => {
      const rl = document.getElementById('rotate-lock');
      rl.getAnimations().forEach(a => { a.pause(); a.currentTime = 0; });
      document.querySelectorAll('#rotate-lock *').forEach(el =>
        el.getAnimations().forEach(a => { a.pause(); a.currentTime = 0; }));
    });
    await p.waitForTimeout(80);
    await p.screenshot({ path: D + 'paysage.png' });
    fs.writeFileSync(D + 'paysage.json', JSON.stringify({ echelle:S, bord:e.bord }));
    const [dans, hors] = execSync('python3 ' + __dirname + '/paysage.py').toString().trim().split(' ').map(Number);

    /* On redresse : l'app doit revenir sans attendre. */
    await p.setViewportSize({ width:w, height:h });
    await p.waitForTimeout(120);
    const retour = await p.evaluate(() => getComputedStyle(document.getElementById('app')).visibility);

    const mauvais = !e.voile || e.vis !== 'hidden' || dans > Math.max(8, hors + 4) || retour !== 'visible';
    if (mauvais) ok = false;
    console.log('  ' + nom.padEnd(17) + (e.voile ? 'levé ' : 'ABSENT').padStart(6) +
      e.vis.padStart(14) + (dans.toFixed(1) + ' %').padStart(18) +
      (hors.toFixed(1) + ' %').padStart(18) + retour.padStart(9) +
      (mauvais ? '   <-- ' + (!e.voile ? 'PAS DE VOILE' : e.vis !== 'hidden' ? 'APP VISIBLE COUCHÉE'
        : retour !== 'visible' ? 'APP RESTÉE INVISIBLE' : 'ÉCRAN COUPÉ EN DEUX') : ''));
    await ctx.close();
  }
  await nav.close();
  console.log(ok ? '\n  OK' : '\n  ÉCHEC');
  process.exit(ok ? 0 : 1);
})();

File: banc-essai/paysage.py
# Cherche UNE COUTURE VERTICALE dans l'image : une colonne de pixels ou la
# couleur saute, sur beaucoup de lignes a la fois. C'est la signature de
# « l'ecran coupe en deux » — l'app figee a 393 px collee a gauche d'un ecran
# de 852, dont le bord droit tranche la photo du haut jusqu'en bas.
# On rend le pourcentage de lignes qui sautent a la colonne la plus marquee,
# d'abord DANS la bande du bord de l'app, puis AILLEURS (temoin : la photo a
# elle seule contient toujours quelques bords verticaux).
import json, sys
from PIL import Image
D = '/tmp/claude-0/-home-user-BibleTrivia/fb9bf869-826b-5523-9825-ea1b24c294d0/scratchpad/'
m = json.load(open(D + 'paysage.json'))
S = m['echelle']            # pixels physiques par pixel CSS
bord = m['bord'] * S        # bord droit de la colonne, en pixels physiques
BANDE = 18 * S              # on cherche a +/- 18 px CSS de ce bord
SEUIL = 24                  # somme des trois canaux : en dessous, c'est du bruit

im = Image.open(D + 'paysage.png').convert('RGB')
W, H = im.size
px = im.load()
lignes = list(range(0, H, 4))

def sauts(x):
    n = 0
    for y in lignes:
        a, b = px[x-1, y], px[x, y]
        if abs(a[0]-b[0]) + abs(a[1]-b[1]) + abs(a[2]-b[2]) > SEUIL: n += 1
    return n

dans, hors = 0, 0
for x in range(1, W):
    n = sauts(x)
    if abs(x - bord) <= BANDE: dans = max(dans, n)
    else:                      hors = max(hors, n)

print('%.1f %.1f' % (100.0*dans/len(lignes), 100.0*hors/len(lignes)))
